crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -27,62 +28,117 @@ use std::sync::Arc;
 /// constraint is applied to that definition, so we'd just return `None`.
 pub(crate) fn narrowing_constraint<'db>(
     db: &'db dyn Db,
-    test: Expression<'db>,
+    constraint: Constraint<'db>,
     definition: Definition<'db>,
 ) -> Option<Type<'db>> {
-    all_narrowing_constraints(db, test)
-        .get(&definition.symbol(db))
-        .copied()
+    match constraint {
+        Constraint::Expression(expression) => all_narrowing_constraints(db, expression)
```

Comment:
I don't think expressions need to be privileged in the naming here, we can rename to `all_narrowing_constraints_for_expression` for better parity.